packages/compose/src/hooks/use-instance-id/index.js

```diff
@@ -0,0 +1,54 @@
+/**
+ * WordPress dependencies
+ */
+import { useMemo, useEffect } from '@wordpress/element';
+
+/**
+ * Array to keep track of allocated ids.
+ */
+const allocatedIds = [];
+
+/**
+ * Find an unallocated id.
+ */
+function findId() {
+	let id = allocatedIds.length;
+	let index = 0;
+
+	while ( index < id ) {
+		if ( ! allocatedIds[ index ] ) {
+			id = index;
+			break;
+		}
+
+		index++;
+	}
+
+	// Allocated the id.
+	allocatedIds[ id ] = true;
+
+	return id;
+}
+
+/**
+ * Free an allocated id.
+ *
+ * @param {number} id Id to free.
+ */
+function freeId( id ) {
+	delete allocatedIds[ id ];
+}
+
+/**
+ * Provides a unique instance ID.
+ */
+export default function useInstanceId() {
+	// Take advantage of useMemo to get the same id throughout the life of a
+	// component.
+	const id = useMemo( findId, [] );
```

Comment:
Seems like this might be a use-case for `useRef` instead?

>The returned object will persist for the full lifetime of the component.

> It’s handy for keeping any mutable value around similar to how you’d use instance fields in classes.

https://reactjs.org/docs/hooks-reference.html#useref

i.e.

```js
const id = useRef( findId() );
// ...
return id.current;
```

Unclear whether it would also make it okay to avoid passing `id` as the dependency of `useEffect` when used this way, if we have a stronger guarantee that the value won't change.

Reply:
`id` itself would never change. `id.current` would. But we should still pass `id`, to align with the planned linting rules.

Reply:
It’s ok if the is changes. With useRef, wouldn’t findId be called on every render?

Reply:
Oh, right, didn't pay attention to that part 😬

Reply:
> With useRef, wouldn’t findId be called on every render?

Yes, you're right 👍 At a glance, I don't see a way around that, so maybe `useMemo` is the way to go.

Reply:
You would need: `useRef( useMemo( findId, [] ) );` hahaha.

Reply:
Actually, using `useMemo` for this is not recommended, because it's not always respected:

https://reactjs.org/docs/hooks-faq.html?no-cache=1#how-to-create-expensive-objects-lazily

Reply:
I know, but as I said before, it's fine if it's not respected. Then a now id is given and the previous id is freed up. It doesn't do any harm, right?

Reply:
Yeah, doesn't do much unless it happens often.

It could be a problem with async mode.

Reply:
Let's see. If need be, we can create an alternative to `useMemo` that never gets called again.

Reply:
Sounds good, a `useStrictMemo` hook 😆